web/src/test/java/org/springframework/security/web/firewall/StrictHttpFirewallTests.java

```diff
@@ -595,4 +597,121 @@ public void getFirewalledRequestWhenUntrustedDomainThenException() {
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getHeader("bad name");
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestGetHeaderWhenNotAllowedHeaderValueThenException() {
+		this.request.addHeader("good name", "bad value");
+		this.firewall.setAllowedHeaderValues(value -> !value.equals("bad value"));
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getHeader("good name");
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestGetDateHeaderWhenControlCharacterInHeaderNameThenException() {
+		this.request.addHeader("Bad\0Name", "some value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getDateHeader("Bad\0Name");
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestGetIntHeaderWhenControlCharacterInHeaderNameThenException() {
+		this.request.addHeader("Bad\0Name", "some value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getIntHeader("Bad\0Name");
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestGetHeaderWhenControlCharacterInHeaderNameThenException() {
+		this.request.addHeader("Bad\0Name", "some value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getHeader("Bad\0Name");
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestGetHeadersWhenControlCharacterInHeaderNameThenException() {
+		this.request.addHeader("Bad\0Name", "some value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getHeaders("Bad\0Name");
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestGetHeaderNamesWhenControlCharacterInHeaderNameThenException() {
+		this.request.addHeader("Bad\0Name", "some value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getHeaderNames().nextElement();
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestGetHeaderWhenControlCharacterInHeaderValueThenException() {
+		this.request.addHeader("Something", "bad\0value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getHeader("Something");
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestGetHeadersWhenControlCharacterInHeaderValueThenException() {
+		this.request.addHeader("Something", "bad\0value");
+
+		HttpServletRequest request = this.firewall.getFirewalledRequest(this.request);
+		request.getHeaders("Something").nextElement();
+	}
+
+	@Test(expected = RequestRejectedException.class)
+	public void getFirewalledRequestGetParameterWhenControlCharacterInParameterNameThenException() {
```

Comment:
Thank you for these tests - will you please add some additional testing around the regular expression? Specifically, I think there should be a test that shows that the firewall requests a header name, a header value, and a parameter name with a control character and with an undefined character.

Reply:
I already had tests for control characters in a header name, a header value, and a parameter name; I added tests for undefined characters for those 3 cases as well.